src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.cs

```diff
@@ -14,6 +18,69 @@ internal static void VerifyValidPath(string path, string argName)
             }
         }
 
+        internal static bool Copy(string sourcePath, string destinationPath, bool recursive,
+            bool skipExistingFiles, CancellationToken cancellationToken)
+        {
+            bool success = true;
+            if (!DirectoryExists(sourcePath))
+            {
+                throw new DirectoryNotFoundException(sourcePath);
+            }
+
+            // Create destination directory if not exists
+            if (!DirectoryExists(destinationPath))
+            {
+                CreateDirectory(destinationPath);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            var fse = new FileSystemEnumerable<(string childPath, bool isDirectory)>(sourcePath,
+                static (ref FileSystemEntry entry) => (entry.ToFullPath(), entry.IsDirectory),
+                recursive ? EnumerationOptions.CompatibleRecursive : EnumerationOptions.Compatible);
+
+            foreach ((string childPath, bool isDirectory) in fse)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                try
+                {
+                    string basePath = childPath[sourcePath.Length..];
+                    if (PathInternal.IsDirectorySeparator(basePath[0]))
+                    {
+                        basePath = basePath[1..];
+                    }
+
+                    string destFilePath = Path.Join(destinationPath, basePath);
+                    if (isDirectory && !DirectoryExists(destFilePath))
+                    {
+                        Directory.CreateDirectory(destFilePath);
+                    }
+                    else
+                    {
+                        // If parent directory was not created in previous case, create it
+                        DirectoryInfo? parentDir = Directory.GetParent(destFilePath);
+                        if (parentDir is { Exists: false })
+                        {
+                            CreateDirectory(parentDir.FullName);
+                        }
+
+                        // Don't copy if file already exists
+                        if (skipExistingFiles && File.Exists(destFilePath))
+                        {
+                            continue;
+                        }
+
+                        CopyFile(childPath, destFilePath, true);
+                    }
+                }
+                catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
+                {
+                    success = false;
+                }
```

Comment:
Does UnauthorizedAccessException can occur for both, source and destination? If so you will need to distinguish between both cases as one is a read error and the other a write one.
Do not ignore all `IOException`s, there are many that are unrelated to read errors which you need to re-`throw`. 
I suggest you to find all the possible "read errors" and how to identify them, then just re-throw for anything else. e.g:
```cs
catch (Exception ex)
{
    if (FileNotFoundError(ex) ||
        DirectorynotFoundError(ex) ||
        SourceFileUnauthorizedAccessError(ex))
    {
        success = false;
    }
    else
    {
        throw ex;
    }
}
```

Some read errors that I can think of are "file not found", "directory not found", unauthorized (only on source file). 
You could check how `CopyFile` does it. You may (or may not) need to make OS-specific implementation of this.

Reply:
The goal of this API was to be simple and at any point someone needs more control over their copy scenario, they should write their own loop. 
@Symbai thoughts?